libraries/abstractions/common_io/test/test_iot_uart.c

```diff
@@ -102,11 +104,17 @@ static IotUARTConfig_t xSampleConfig2 =
 static void prvReadWriteCallback( IotUARTOperationStatus_t xOpStatus,
                                   void * pvParams )
 {
-    IotUARTOperationStatus_t * xCallbackStatus = ( IotUARTOperationStatus_t * ) pvParams;
     BaseType_t xHigherPriorityTaskWoken;
 
-    *xCallbackStatus = xOpStatus;
-    xSemaphoreGiveFromISR( xtestIotUARTSemaphore, &xHigherPriorityTaskWoken );
+    if( xOpStatus == eUartReadCompleted )
```

Comment:
This branch suggests that it might be cleaner to refactor this into two separate callbacks.

However, for example, what if the read operation has ended, but it failed? Will the same xOpStatus value be returned in that case?

Reply:
In current API design, there is only one callback registered for both read and write. So I think it is not easy to do 2 separate callbacks for async_read and async_write.

https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/common_io/include/iot_uart.h#L205

If the read failed, it will be different error code in which case the semaphore will not be given. Then the test case that waits for the semaphore will fail. So I think this flow should work well.